analyzer/src/main/resources/scripts/init.gradle

```diff
@@ -138,22 +138,21 @@ class AbstractDependencyTreePlugin<T> implements Plugin<T> {
     }
 
     private static class DependencyTreeModelBuilder implements ToolingModelBuilder {
+        /**
+         * Stores the Dependency objects created by this builder, so that they can be reused when the same
+         * dependency is encountered again in the dependency graph. As dependencies can occur many times in large
+         * dependency graphs, de-duplicating these objects can save a significant amount of memory.
+         *
+         * Note, however, that packages can be referenced in the graph multiple times with a different set of
+         * dependencies. Therefore, for a package with a given identifier, there can be different Dependency
+         * objects; hence the values of the map are lists.
+         */
+        private final Map<String, List<Dependency>> dependencies = new HashMap<>()
+
         static DependencyImpl dependencyFromDisplayName(String displayName, List<Dependency> dependencies,
                                                         String error, String warning) {
-            if (displayName.startsWith('project :')) {
-                def coordinates = displayName.split(':', 2)
-                return new DependencyImpl(groupId: '<project>', artifactId: coordinates[1], dependencies: dependencies,
-                        error: error?.toString(), warning: warning?.toString())
-            }
-
-            def coordinates = displayName.split(':')
-
-            if (coordinates.length == 3) {
-                return new DependencyImpl(groupId: coordinates[0], artifactId: coordinates[1], version: coordinates[2],
-                        dependencies: dependencies, error: error?.toString(), warning: warning?.toString())
-            }
-
-            return new DependencyImpl(groupId: '<unknown>', artifactId: displayName.replace(':', '_'),
+            def coordinates = displayNameToCoordinates(displayName)
```

Comment:
Could we maybe split out changes (like this refactoring) that are not strictly related to the main topic to preceding commits? This make the diff for the important change smaller and thus easier to review.

(This might also be applicable to other functions below, but you know better than me whether that's the case.)

Reply:
Created a separate commit for the refactoring of this function. The other functions belong to the logic to de-duplicate dependencies; so they are left in the main commit.